Standards/scs-0217-v1-baseline-cluster-security.md

```diff
@@ -0,0 +1,145 @@
+specific functions of this API. This is the case for users as well as internal components.
+
+Most internal clients (like proxies or nodes) are typically authenticated via service accounts or
+x509 certificates, which will normally be created automatically during the setup of a cluster.
+External users can authenticate via an access pattern of choice, which is typically decided by
+the cluster provider.
+
+Authorization is (normally) done by the Role-Based Access Control (RBAC), which matches a request
+by a user with a set of permissions, also called a role. Kubernetes deploys some roles out-of-the-box;
+additional roles need to be carefully checked, since some permissions for specific resources allow
+modification of other resources.
+
+This whole process is especially important for the Kubelet, which allows anonymous requests in its
+default configuration. This is obviously a security risk, since everybody with access to its endpoint
+could manipulate resources that are managed with the Kubelet.
+
+To disable anonymous requests, the Kubelet should be started with `--anonymous-auth=false`.
+Authentication can be provided either through x509 client certificates or API bearer tokens.
+How to set up both approaches can be found in the [Kubelet Authentication and Authorization](https://kubernetes.io/docs/reference/access-authn-authz/kubelet-authn-authz/).
+
+Kubelet authorization is set to `AlwaysAllow` as a default mode. This can be quite problematic,
+since all authenticated users can do all actions. To mitigate this, it is possible to delegate
+authorization to the API server by:
+
+- enabling the `authorization.k8s.io/v1beta1` API group
+- starting the Kubelet with the `--authorization-mode=Webhook` and the `--kubeconfig` flags
+
+After that, the Kubelet calls the `SubjectAccessReview` API in order to determine the authorization of a request.
+
+## Decision
+
+This standard tries to increase security for a Kubernetes cluster in order to provide a
+solid baseline setup with regard to security. For this to work, multiple measures need to be undertaken.
+
+A self-controlled CA SHOULD be used in order to be in control of the TLS certificates, which
+enables the operator to provide and revoke certificates according to the requirements.
+
+All internal endpoints found in the section [TLS for all internal/API traffic] MUST be
+encrypted with TLS in order to secure internal traffic.
+
+The Kubernetes API (kubeAPI) MUST be secured by authenticating and authorizing the users
+trying to access its endpoints. How a user is authenticated is up to the provider of the
+cluster and/or the wishes of the customer. Authorization MUST be done by providing
+fine-grained RBAC.
+The authentication and authorization steps MUST also be applied to the Kubelet, which in its default configuration
+doesn't enable them. A way to do this can be found in the section [Authentication and Authorization].
```

Comment:
I find this very unspecific for a standards document.

Especially with regards to security this allows for many possible implementations which usually leads to - inadvertently - insecure implementations because there is no clear mandatory way to configure things in a secure fashion.

this is what opened many security holes in previous TLS standards, for example.

See this RFC for recommendations on how to write security related standards documents:

https://www.rfc-editor.org/rfc/rfc3552.html#page-26

I want to particularly highlight something relevant to the above:

> If the technology involves authentication, particularly user-host
   authentication, the security of the authentication method MUST be
   clearly specified.  That is, authors MUST document the assumptions
   that the security of this authentication method is predicated upon.

So, to sum it up:

If you don't enforce secure authentication and authorization by the standard it's a matter of time until someone implementing your standard will end up implementing it in an insecure manner.

It is also not sufficient that the standard relies on third party security-by-accident schemes, made up example:

"the supported authentication schemes in kubernetes are all secure, therefore we do not need to specify explicitly secure auth schemes".

A standard should not rely on implementation specific behaviour, which might change in the future, or might be version specific.

Be explicit what is secure and be strict, not liberal in what you accept, see this famous criticism of the robustness principle:

https://datatracker.ietf.org/doc/html/draft-iab-protocol-maintenance

HTH

Reply:
Why was this marked as "resolved" @jschoone ? There wasn't even any effort done to reply to this review in any form.

If you disagree with the stated observations and opinions in a review please state that in written form. Why do you think this is not an issue or why it does not matter, or how this is already addressed in the standard or elsewhere?

Just marking stuff as "resolved" in order to unblock work devalues the work of the reviewer and is a really bad practice imho.

Please don't do that anymore, at least for my reviews.

Reply:
marking this as "resolved" as I don't want to block you.